File: astrbot/dashboard/api/coding_cli.py
"""Manage the provider presets a coding CLI's own configuration is switched between.

The run path layers a config the CLI loads in addition to the user's own, so a
delegated run never rewrites what the operator runs interactively.  These routes
are the other, explicitly-invoked path: the operator asking for one provider to
become the CLI's default on this host.  That replaces the user's own settings
and persists a credential in a file AstrBot does not own, so a switch is
high-risk and therefore step-up.

The provider list itself is ordinary configuration (``btw.cli_providers``), and
is saved through the config profile the same way any other setting is.  Only
the switch touches a file outside AstrBot's data directory.
"""

from typing import Any, NoReturn

from fastapi import APIRouter, Depends, Request

from astrbot.core.agent.btw import cli_global_config
from astrbot.core.auth.models import Resource
from astrbot.dashboard.responses import ApiError, ok
from astrbot.dashboard.schemas import CodingCliSwitchRequest

from .auth import (
    AuthContext,
    require_resource_action,
    require_scope,
)

router = APIRouter(tags=["Coding CLI"])

# Every write into a CLI's own global config asks for this one action.
WRITE_ACTION = "coding_cli.config.write"


async def require_coding_cli_scope(request: Request) -> AuthContext:
    """Authenticate as config management; the action is decided per resource."""
    return await require_scope(request, "config", authorize_action=False)


async def authorize_coding_cli(request: Request, auth: AuthContext) -> None:
    """Require the high-risk action that switching a CLI's provider needs."""
    await require_resource_action(
        request,
        auth,
        action=WRITE_ACTION,
        resource=Resource.instance("default"),
    )


def _raise(exc: Exception) -> NoReturn:
    raise ApiError(str(exc), status_code=400) from exc


def _state(cli: str) -> dict[str, Any]:
    """Serialize one CLI's state without ever naming a stored credential."""
    state = cli_global_config.read_state(cli)
    return {
        "cli": cli,
        "path": str(state.path),
        "exists": state.exists,
        "managed": state.managed,
        "backed_up": state.backed_up,
        "base_url": state.base_url,
        "model": state.model,
        "has_credential": state.has_credential,
    }


def _is_current(provider: dict, state: cli_global_config.ManagedFile) -> bool:
    """Return whether the CLI's file currently holds this provider.

    Compares the endpoint and model rather than the credential: AstrBot cannot
    read a stored key back, and the endpoint is what identifies a provider.
    """
    if not state.managed or not provider["base_url"]:
        return False
    if provider["base_url"] != state.base_url:
        return False
    return not provider["model"] or provider["model"] == state.model


def _cli_payload(cli: str, raw: Any) -> dict[str, Any]:
    """One CLI's file state plus the providers it can be switched to."""
    state = cli_global_config.read_state(cli)
    return {
        **_state(cli),
        "providers": [
            {
                "id": provider["id"],
                "name": provider["name"],
                "base_url": provider["base_url"],
                "model": provider["model"],
                "note": provider["note"],
                # A key is reported as present, never as its value.
                "has_api_key": bool(provider["api_key"]),
                "current": _is_current(provider, state),
            }
            for provider in cli_global_config.providers_for_cli(raw, cli)
        ],
    }


def _provider_for(cli: str, config: Any, provider_id: str) -> dict:
    """Return the named provider from the profile's list, or refuse."""
    btw = config.get("btw", {}) if isinstance(config, dict) else {}
    raw = btw.get("cli_providers") if isinstance(btw, dict) else None
    provider = cli_global_config.select_provider(raw, provider_id)
    if provider is None:
        raise ApiError(f"No provider with the id {provider_id!r} is configured.")
    if provider["cli"] and provider["cli"] != cli:
        raise ApiError(
            f"The provider {provider_id!r} is scoped to {provider['cli']!r}, so it "
            f"cannot be applied to {cli!r}."
        )
    return provider


def _config_of(request: Request) -> Any:
    """Return the running profile, which holds the operator's provider list."""
    runtime = getattr(request.app.state, "runtime", None)
    services = getattr(runtime, "services", None)
    profiles = getattr(services, "config_profiles", None)
    acm = getattr(profiles, "acm", None)
    configs = getattr(acm, "confs", None)
    if not isinstance(configs, dict):
        raise ApiError("Configuration is unavailable", status_code=503)
    return configs.get("default", {})


@router.get("/coding-cli/global-config")
async def get_global_config(
    request: Request,
    auth: AuthContext = Depends(require_coding_cli_scope),
):
    """Report each CLI's own configuration and the providers it can use."""
    await authorize_coding_cli(request, auth)
    btw = _config_of(request).get("btw", {})
    raw = btw.get("cli_providers") if isinstance(btw, dict) else None
    return ok(
        {
            "clis": [
                _cli_payload(cli, raw)
                for cli in cli_global_config.CODING_GLOBAL_AGENT_TYPES
            ]
        }
    )


@router.put("/coding-cli/global-config")
async def switch_provider(
    payload: CodingCliSwitchRequest,
    request: Request,
    auth: AuthContext = Depends(require_coding_cli_scope),
):
    """Make one configured provider the CLI's default on this host."""
    await authorize_coding_cli(request, auth)
    provider = _provider_for(payload.cli, _config_of(request), payload.provider_id)
    try:
        cli_global_config.apply_provider(payload.cli, provider)
    except (OSError, ValueError) as exc:
        _raise(exc)
    return ok(_state(payload.cli))


@router.delete("/coding-cli/global-config/{cli}")
async def remove_global_config(
    cli: str,
    request: Request,
    auth: AuthContext = Depends(require_coding_cli_scope),
):
    """Take AstrBot's configuration back out of the CLI's own file."""
    await authorize_coding_cli(request, auth)
    if cli not in cli_global_config.CODING_GLOBAL_AGENT_TYPES:
        raise ApiError(f"Unknown coding agent type: {cli!r}")
    try:
        cli_global_config.remove_provider(cli)
    except (OSError, ValueError) as exc:
        _raise(exc)
    return ok(_state(cli))

File: dashboard/src/api/v1/codingCli.ts
import { generatedOptions, openApiV1, typed } from './shared';
import type { AxiosRequestConfig } from './shared';

/** The CLIs AstrBot knows how to write a global configuration for. */
export type CodingCliKind = 'claude_code' | 'codex';

/** One provider the operator configured for a CLI. */
export interface CodingCliProvider {
  id: string;
  name: string;
  base_url: string;
  model: string;
  note: string;
  /** A key is reported as present, never as its value. */
  has_api_key: boolean;
  /** Whether the CLI's own file currently holds this provider. */
  current: boolean;
}

/** One CLI's own configuration file, plus what it can be switched to. */
export interface CodingCliState {
  cli: CodingCliKind;
  path: string;
  exists: boolean;
  managed: boolean;
  backed_up: boolean;
  base_url: string;
  model: string;
  has_credential: boolean;
  providers: CodingCliProvider[];
}

/**
 * One provider as `btw.cli_providers` stores it.
 *
 * `api_key` is what the operator typed in the field and nothing else.  A config
 * response writes the marker `__ASTRBOT_REDACTED__` where a stored key would
 * be, so the editor treats it as "a key is stored, the field is empty": the
 * marker goes back only for the entry it came from, and only a switch ever
 * writes a value into the CLI's own file.
 */
export interface StoredCliProvider extends CodingCliProvider {
  /** The CLI this provider is scoped to; empty means either CLI may use it. */
  cli: string;
  api_key: string;
}

export interface CodingCliStatePayload {
  clis: CodingCliState[];
}

export const codingCliApi = {
  /**
   * Read each CLI's own configuration.
   *
   * A request config is optional and carries the step-up credential when the
   * backend asks for one: the read is an ordinary permission, but an operator
   * whose session has not proved itself yet still has to answer the challenge
   * before the list can be filled in.
   */
  state(requestConfig?: AxiosRequestConfig) {
    return typed<CodingCliStatePayload>(
      openApiV1.getCodingCliGlobalConfig(generatedOptions({}, requestConfig)),
    );
  },
  switchProvider(
    payload: { cli: CodingCliKind; provider_id: string },
    requestConfig?: AxiosRequestConfig,
  ) {
    return typed<CodingCliState>(
      openApiV1.switchCodingCliProvider(
        generatedOptions(
          {
            body: { cli: payload.cli, provider_id: payload.provider_id },
          },
          requestConfig,
        ),
      ),
    );
  },
  remove(cli: CodingCliKind, requestConfig?: AxiosRequestConfig) {
    return typed<CodingCliState>(
      openApiV1.removeCodingCliGlobalConfig(
        generatedOptions({ path: { cli } }, requestConfig),
      ),
    );
  },
};
